File: src/main/java/com/example/dpop/orchestrator/api/v1/authentication/sms/SmsEnrollPending.java
package com.example.dpop.orchestrator.api.v1.authentication.sms;

import com.example.dpop.auth_sms.EnrollmentRef;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public record SmsEnrollPending(
        String phoneNumber,
        EnrollmentRef enrollmentRef,
        String tan,
        boolean tanVerified,
        boolean enrollmentConfirmed) {

    public static SmsEnrollPending empty() {
        return new SmsEnrollPending(null, null, null, false, false);
    }

    public SmsEnrollPending merge(Map<String, Object> patch) {
        if (patch == null) return this;
        String phone = patch.containsKey("phoneNumber") ? (String) patch.get("phoneNumber") : phoneNumber;
        String newTan = patch.containsKey("tan") ? (String) patch.get("tan") : tan;
        return new SmsEnrollPending(phone, enrollmentRef, newTan, tanVerified, enrollmentConfirmed);
    }

    public SmsEnrollPending withEnrollmentRef(EnrollmentRef ref) {
        return new SmsEnrollPending(phoneNumber, ref, tan, false, false);
    }

    public SmsEnrollPending withTanVerified() {
        return new SmsEnrollPending(phoneNumber, enrollmentRef, tan, true, false);
    }

    public SmsEnrollPending withEnrollmentConfirmed() {
        return new SmsEnrollPending(phoneNumber, enrollmentRef, tan, true, true);
    }

    public List<String> missingUserInputs() {
        List<String> missing = new ArrayList<>();
        if (phoneNumber == null || phoneNumber.isBlank()) missing.add("phoneNumber");
        if (tan == null || tan.isBlank()) missing.add("tan");
        return missing;
    }
}

File: src/main/kotlin/com/example/dpop/orchestrator/api/v1/authentication/sms/SmsEnrollPending.kt
package com.example.dpop.orchestrator.api.v1.authentication.sms

import com.example.dpop.auth_sms.EnrollmentRef

@JvmRecord
data class SmsEnrollPending(
    val phoneNumber: String?,
    val enrollmentRef: EnrollmentRef?,
    val tan: String?,
    val tanVerified: Boolean,
    val enrollmentConfirmed: Boolean
) {

    fun merge(patch: Map<String, Any>?): SmsEnrollPending {
        if (patch == null) return this
        val phone = if (patch.containsKey("phoneNumber")) patch["phoneNumber"] as String? else phoneNumber
        val newTan = if (patch.containsKey("tan")) patch["tan"] as String? else tan
        return SmsEnrollPending(phone, enrollmentRef, newTan, tanVerified, enrollmentConfirmed)
    }

    fun withEnrollmentRef(ref: EnrollmentRef?): SmsEnrollPending =
        SmsEnrollPending(phoneNumber, ref, tan, false, false)

    fun withTanVerified(): SmsEnrollPending =
        SmsEnrollPending(phoneNumber, enrollmentRef, tan, true, false)

    fun withEnrollmentConfirmed(): SmsEnrollPending =
        SmsEnrollPending(phoneNumber, enrollmentRef, tan, true, true)

    fun missingUserInputs(): List<String> {
        val missing = mutableListOf<String>()
        if (phoneNumber.isNullOrBlank()) missing.add("phoneNumber")
        if (tan.isNullOrBlank()) missing.add("tan")
        return missing
    }

    companion object {
        @JvmStatic
        fun empty(): SmsEnrollPending = SmsEnrollPending(null, null, null, false, false)
    }
}
